src/Fantomas.Core/MultipleDefineCombinations.fs

```diff
@@ -0,0 +1,284 @@
+///     non existing content (for a specific combination of defines) or
+///     active content.
+///
+/// When the code need to be compared, a CustomComparison is used to determine which fragment we are interested in.
+[<RequireQualifiedAccess>]
+[<CustomEquality; CustomComparison>]
+type CodeFragment =
+    /// Any line that starts with `#if`, `#else` or `#endif`
+    | HashLine of line: string * defines: DefineCombination
+    /// Content found between two HashLines
+    | Content of code: string * lineCount: int * defines: DefineCombination
+    /// When two HashLine follow each other without any content in between.
+    | NoContent of defines: DefineCombination
+
+    member x.Defines =
+        match x with
+        | HashLine(defines = defines)
+        | Content(defines = defines)
+        | NoContent(defines = defines) -> defines
+
+    member x.LineCount =
+        match x with
+        | HashLine _ -> 1
+        | Content(lineCount = lineCount) -> lineCount
+        | NoContent _ -> 0
+
+    override this.Equals _ = false
+    override this.GetHashCode() = Int32.MinValue
+
+    interface IComparable with
+        member x.CompareTo other =
+            match other with
+            | :? CodeFragment as other -> (x :> IComparable<_>).CompareTo other
+            | _ -> -1
+
+    interface IComparable<CodeFragment> with
+        member x.CompareTo y =
+            match x, y with
+            // When comparing the different results of each format result, the single constant is that all hash lines
+            // should exactly match.
+            | CodeFragment.HashLine(line = lineX), CodeFragment.HashLine(line = lineY) ->
+                assert (lineX = lineY)
+                0
+            // Pick the other fragment is it has content you don't
+            | CodeFragment.NoContent _, CodeFragment.Content _ -> -1
+            // Pick our fragment is the other fragment has no code
+            | CodeFragment.Content _, CodeFragment.NoContent _ -> 1
+            // If both fragments are empty they are equivalent.
+            // Keep in mind that we could be comparing more the two fragments at the same time in `traverseFragments`
+            | CodeFragment.NoContent _, CodeFragment.NoContent _ -> 0
+            // If both fragments have content, we want to take the content with the most lines.
+            | CodeFragment.Content(lineCount = ownLineCount; code = ownContent),
+              CodeFragment.Content(lineCount = otherLineCount; code = otherContent) ->
+                let hasOwnContent = not (String.IsNullOrWhiteSpace ownContent)
+                let hasOtherContent = not (String.IsNullOrWhiteSpace otherContent)
+
+                if ownLineCount > otherLineCount then
+                    1
+                elif ownLineCount < otherLineCount then
+                    -1
+                elif
+                    String.IsNullOrWhiteSpace ownContent
+                    && not (String.IsNullOrWhiteSpace otherContent)
```

Comment:
Isn't this the same as line 79?: `elif not hasOwnContent && hasOtherContent then`